requre/cassette.py

```diff
@@ -19,16 +19,716 @@
+    """
 
     @property
     def content(self) -> dict:
-        return PersistentObjectStorage().storage_object
+        return self.storage_object
+
+    internal_object_key = METATADA_KEY
+    version_key = "version_storage_file"
+    key_inspect_strategy_key = "key_strategy"
+
+    def _set_defaults(self) -> None:
+        self.dump_after_store = False
+        storage_mode = os.getenv(ENV_REQURE_STORAGE_MODE, "default")
+        logger.info(
+            f"You overrided storage mode via env var {ENV_REQURE_STORAGE_MODE}={storage_mode}"
+        )
+        if not hasattr(StorageMode, storage_mode):
+            raise PersistentStorageException(
+                f"storage mode '{storage_mode}' does not exist, "
+                f"use one of {list(StorageMode.__members__.keys())}) "
+            )
+        self.mode: StorageMode = getattr(StorageMode, storage_mode)
+        self.is_flushed = False
+        self.storage_object: dict = {}
+        self._storage_file: Optional[str] = None
+        self.data_miner = DataMiner()
+
+    def __init__(self) -> None:
+        # call dump() after store() is called
+        self._set_defaults()
+        storage_file_from_env = os.getenv(ENV_STORAGE_FILE)
+        if storage_file_from_env:
+            self.storage_file = storage_file_from_env
+
+    def __str__(self) -> str:
+        return (
+            f"{self.__class__.__name__}("
+            f"storage_mode={self.mode}, "
+            f"storage_file={self._storage_file})"
+        )
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+    @property
+    def metadata(self):
+        """
+        Placeholder to store some custom configuration, or store custom data,
+        eg. version of storage file, or some versions of packages
+
+        :return: dict
+        """
+        return self.storage_object.get(self.internal_object_key, {})
+
+    @metadata.setter
+    def metadata(self, key_dict: dict):
+        if self.internal_object_key not in self.storage_object:
+            self.storage_object[self.internal_object_key] = {}
+        for k, v in key_dict.items():
+            self.storage_object[self.internal_object_key][k] = v
+        self.is_flushed = False
+
+    @property
+    def storage_file_version(self):
+        """
+        Get version of persistent storage file
+        :return: int
+        """
+        return self.metadata.get(self.version_key, 0)
+
+    def do_store(self, keys):
+        if self.mode == StorageMode.read and not os.path.exists(self.storage_file):
+            raise PersistentStorageException(
+                "Requre can't work in this setup: we are meant to read "
+                "recorded responses but the storage file does not exist."
+            )
+        if self.mode == StorageMode.read:
+            return False
+
+        elif self.mode == StorageMode.write:
+            return True
+
+        elif self.mode == StorageMode.append:
```

Comment:
We need to think about a better approach for those `if-else` for each mode. (Not here..;) (It's present in multiple places.)

I would use a class approach with the well-defined interface for all those strategy-like types like mode, data_type, key_strategy,... It will make the code more readable and allows us to create new implementations easily. (Without forgetting to fill some `if-else`.)